### PR TITLE
WIP Add AtomicFolder

### LINKED ISSUE #35: Add `atomic_folder`
Sometimes I have a bunch of operations which create many files in a folder. The folder should only be there in the end of the operation. Something like this:
```python
with atomic_folder("foo") as folder:
    unzip(something, folder)
    # "foo/" doesn't exist yet.
# now it does
```

### PR DESCRIPTION
**Pull request by sotte**
This is pretty much just copy&paste from one of my projects. Your version is much more sophisticated but I think it's good to have the code just to get the discussion started.

Things that have to be addressed:
- [ ] `rename_dont_move` could be replaced by existing functionality.
- [ ] Test on windows.
- [ ] Adjust coding style
- [ ] Update docs
- [ ] Write tests
- [ ] Anything else?

@untitaker what do you think?

I'll clean up the commit history once the PR is ready.

This PR closes #35.

### CODE REVIEW

File: atomicwrites/__init__.py [hunk 1]
```diff
@@ -199,3 +199,68 @@ def atomic_write(path, writer_cls=AtomicWriter, **cls_kwargs):
     :py:class:`AtomicWriter`.
     '''
     return writer_cls(path, **cls_kwargs).open()
+
+
+
+###############################################################################
```
Comment (untitaker) [inline atomicwrites/__init__.py:205]: If this is necessary I think it's time to start a new file! Please create a file `folders.py` and import `AtomicFolder` from it.

File: atomicwrites/__init__.py [hunk 2]
```diff
@@ -199,3 +199,68 @@ def atomic_write(path, writer_cls=AtomicWriter, **cls_kwargs):
...
+
+def rename_dont_move(path, dst):
+    """
+    Rename path to dst only if dst does not exist.
+
+    See http://luigi.readthedocs.io/en/stable/_modules/luigi/target.html#FileSystem.rename_dont_move
+    """
+    if os.path.isdir(dst):
+        raise FolderAlreadyExists("'{}' already exists".format(dst))
+    shutil.move(path, dst)
```
Comment (untitaker) [inline atomicwrites/__init__.py:266]: This function is not atomic as the implementation in luigi notes in its warning message. Here is a race condition for example: We have two processes A and B, which both run this function.

* Process A runs `isdir`, which returns false
* Process B runs `isdir`, which returns false
* Process A moves its tmpdir to the target path
* Process B moves its tmpdir to the target path (or rather, *into* the directory created by A)

I am not sure whether we can guarantee to fail atomically if the target path doesn't exist. If you have any ideas regarding that... Ideally we would have an `overwrite` param but I am not sure if that is possible.

File: atomicwrites/__init__.py [hunk 3]
```diff
@@ -199,3 +199,68 @@ def atomic_write(path, writer_cls=AtomicWriter, **cls_kwargs):
...
+# Atomic folder
+import os
+from os.path import dirname, basename
+import shutil
+
+
+class FolderAlreadyExists(Exception): pass
+
+
+class AtomicFolder():
```
Comment (untitaker) [inline atomicwrites/__init__.py:215]: Please model `AtomicFolder` like the API of `AtomicWriter` for consistency, concretely that it itself is not a contextmanager but rather has a `open()` method that returns a context manager, and separate methods for `commit` and `rollback`.

**Review by untitaker — CHANGES_REQUESTED** [review]
Regarding codestyle, I will eventually run [black](https://github.com/ambv/black) against the entire codebase, so don't worry too much about it.

File: atomicwrites/__init__.py (hunk 3 above)
Comment (untitaker) [reply atomicwrites/__init__.py:215]: Oh and please subclass from `object` such that we have a newstyle class on both Python 2 and 3.

File: atomicwrites/__init__.py (hunk 1 above)
Comment (sotte) [reply atomicwrites/__init__.py:205]: Sure, I'll do that once the content is fine!

**Comment by untitaker** [comment]
See discussion on #35

**Pull request closed without merging**